WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemDataSource.kt

```diff
@@ -0,0 +1,112 @@
+    data class RemotePostId(val id: RemoteId) : PostListItemIdentifier()
+    object EndListIndicatorIdentifier : PostListItemIdentifier()
+}
+
+class PostListItemDataSource(
+    private val dispatcher: Dispatcher,
+    private val postStore: PostStore,
+    private val performGetItemIdsToHide: ((PostListDescriptor) -> Pair<LocalPostId, RemotePostId>?),
+    private val transform: ((PostModel) -> PostListItemUiState)
+) : ListItemDataSourceInterface<PostListDescriptor, PostListItemIdentifier, PostListItemType> {
+    override fun fetchList(listDescriptor: PostListDescriptor, offset: Long) {
+        val fetchPostListPayload = FetchPostListPayload(listDescriptor, offset)
+        dispatcher.dispatch(PostActionBuilder.newFetchPostListAction(fetchPostListPayload))
+    }
+
+    override fun getItemIdentifiers(
+        listDescriptor: PostListDescriptor,
+        remoteItemIds: List<RemoteId>,
+        isListFullyFetched: Boolean
+    ): List<PostListItemIdentifier> {
+        val localItems = postStore.getLocalPostIdsForDescriptor(listDescriptor)
+                .map { LocalPostId(id = it) }
+        val remoteItems = remoteItemIds.map { RemotePostId(id = it) }
+        val actualItems = localItems + remoteItems
+
+        val endListItem = if (isListFullyFetched && actualItems.isNotEmpty()) {
+            listOf(EndListIndicatorIdentifier)
+        } else emptyList()
+        val itemsToHide = getItemIdsToHide(listDescriptor)
+        // We only want to show the end list indicator if the list is fully fetched and it's not empty
+        return (actualItems + endListItem).filter { !itemsToHide.contains(it) }
+    }
+
+    override fun getItemsAndFetchIfNecessary(
+        listDescriptor: PostListDescriptor,
+        itemIdentifiers: List<PostListItemIdentifier>
+    ): List<PostListItemType> {
+        val localOrRemoteIds = itemIdentifiers.mapNotNull { identifier ->
+            when (identifier) {
+                is LocalPostId -> identifier.id
+                is RemotePostId -> identifier.id
+                // We are creating a list of local and remote ids, so other type of identifiers don't matter
+                EndListIndicatorIdentifier -> null
+            }
+        }
+
+        val postList = postStore.getPostsByLocalOrRemotePostIds(localOrRemoteIds, listDescriptor.site)
+
+        // Convert the post list into 2 maps with local and remote ids as keys
+        val localPostMap = postList.associateBy { LocalId(it.id) }
+        val remotePostMap = postList.filter { it.remotePostId != 0L }.associateBy { RemoteId(it.remotePostId) }
+
+        // Fetch the missing posts
+        val remoteIdsToFetch: List<RemoteId> = localOrRemoteIds.mapNotNull { it as? RemoteId }
+                .filter { !remotePostMap.containsKey(it) }
+        fetchPosts(listDescriptor.site, remoteIdsToFetch)
+
+        val transformFromNullablePost = { localOrRemoteId: LocalOrRemoteId, post: PostModel? ->
```

Comment:
Nitpick: I feel like this lambda makes the code just more difficult to understand. How do you feel about moving it into a regular method - we could call it something like "transformToUiState"?

Reply:
Fixed in 093558595838c2a8219b6991cb9e1bb4bbc48e62.